pkg/bpf/collection.go

```diff
@@ -41,6 +45,110 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
 	return spec, nil
 }
 
+func removeUnreachableTailcalls(spec *ebpf.CollectionSpec) error {
+	type TailCall struct {
+		referenced bool
+		visited    bool
+		spec       *ebpf.ProgramSpec
+	}
+
+	entrypoints := make([]*ebpf.ProgramSpec, 0)
+	tailcalls := make(map[uint32]*TailCall)
+
+	for _, prog := range spec.Programs {
+		if prog.SectionName == "tc" || prog.SectionName == "xdp" {
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		var id, slot uint32
+		if _, err := fmt.Sscanf(prog.SectionName, "%d/%v", &id, &slot); err != nil {
+			continue
+		}
+
+		const (
+			cilium_policy_map        = 1
+			cilium_calls_map         = 2
+			cilium_egress_policy_map = 4
+		)
+
+		// Any programs in the cilium_policy_map or cilium_egress_policy_map are
+		// called dynamically, assume they are reachable and treat as entrypoints.
+		if id == cilium_policy_map || id == cilium_egress_policy_map {
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		// Any programs not in the tail call map can be ignored (always load them).
+		if id != cilium_calls_map {
+			continue
+		}
+
+		tailcalls[slot] = &TailCall{
+			spec: prog,
+		}
+	}
+
+	// Discover all tailcalls that are reachable from the given program.
+	visit := func(prog *ebpf.ProgramSpec, tailcalls map[uint32]*TailCall) {
+		for i := 0; i < len(prog.Instructions); i++ {
+			inst := prog.Instructions[i]
+			if !(inst.IsBuiltinCall() && inst.Constant == int64(asm.FnTailCall)) {
+				continue
+			}
+
+			if i-1 <= 0 {
+				continue
+			}
+
+			// For all tail calls where the index is static we use the `tail_call_static` function.
```

Comment:
Does it make sense to document in the C for `tail_call_static` that this function relies on the specific asm?

Reply:
Maybe even add sig-loader to CODEOWNERS for that file.

Reply:
Yea, I think you are right.